### ISSUE TITLE
Can't clone Rc

### ISSUE THREAD

**Issue by jgarvin** [point-rs/shared-rc#1]
Wondering if you could help with deciphering some errors using shared_rc :)

Main question is what is going on below, that makes it so taking code using `std::rc::Rc` and substituting `shared_rc::Rc` breaks compilation. First the working version:

```
thread_local!(
    static WAT: std::rc::Rc<i32> = std::rc::Rc::new(42);
);

fn compiles_fine() -> std::rc::Rc<i32> {
    WAT.with(|wat| wat.clone())
}
```

Then the broken version:

```
thread_local!(
    static WAT: shared_rc::Rc<i32> = shared_rc::Rc::new(42);
);

// Complains that we are returning a `&Rc` instead of an `Rc`?! If you
// use rust analyzer to lookup which `clone` is being called, it's
// calling the blanket impl of `clone` for `&T`. So somehow we are
// cloning a `&Rc` instead of the `Rc`. `with` is passing in a `&Rc`,
// so I guess it thinks we're cloning that. At a minimum though,
// behavior is diverging from `std::rc::Rc` in some weird way, because
// we're doing the exact same thing as in `compiles_fine`.
fn gives_error() -> shared_rc::Rc<i32> {
    WAT.with(|wat| wat.clone())
}

// So lets deref the `&Rc`, which should get us the clone method for
// the `Rc` itself... right? No, now the compiler complains we're
// returning `i32` instead of `Rc`. `Deref` recursing caused us to go
// straight through both layers I guess, although I would have
// expected that to only happen from the implicit `deref`'ing done to
// find the method to call, not an explicit * operator... wat.
fn still_gives_error() -> shared_rc::Rc<i32> {
    WAT.with(|wat| (*wat).clone())
}
```

**Comment by CAD97** [comment]
The issue is that `SharedRc<i32>` isn't `Clone`. Which is because I used `#[derive(Clone)]` instead of a manual `impl Clone`, so the `Clone` impl is conditional on the payload being `Clone`.

Importantly, the full type is `SharedRc<i32, dyn Destruct>`. Because the host type is a trait object, it isn't `Clone`, so the derived `Clone` impl doesn't apply.

If you want to try taking an attempt at patching this, it should be just replacing the `#[derive(Clone)]` with the more permissive `impl<T: ?Sized, Owner: ?Sized> Clone for $Rc<T, Owner>`. I'll hopefully get around to patching and publishing a fix for the silly oversight sometime relatively soonish.

**Comment by jgarvin** [comment]
Taking a crack at this since I'm also interested in tweaking the signature of `project` to return `U + '_`  instead of `&U`. On the rust user's forum you also mentioned this:

> ... And my crate is probably completely unsound at the moment due to improper Send/Sync impl bounds. Oof.

In the code I see:

        unsafe impl<T: ?Sized, Owner: ?Sized> Send for $Rc<T, Owner> where rc::$Rc<T>: Send {}
        unsafe impl<T: ?Sized, Owner: ?Sized> Sync for $Rc<T, Owner> where rc::$Rc<T>: Sync {}

AFAIK we want `Rc` to not be Send or Sync, and we want `Arc` to be both. However, I'm confused by the above for several reasons:

* We don't want any Send/Sync impl for `Rc` at all AFAICT, only `Arc`. Maybe this was thought to be harmless for `Rc` case? Maybe it should be moved out of the macro and just done manually for the `Arc` case?
* This seems to be saying that `Rc<T, Owner>` is Send+Sync if `Rc<T>` is. So `Owner` doesn't affect sendability/syncability, but I don't see why it couldn't? Is that what is improper?
* Why does rustc accept this, since `Rc<T>` is just shorthand for `Rc<T, dyn Destruct>` which this impl would apply to so then it would read `unsafe impl<T: ?Sized> Send for $Rc<T, dyn Destruct> where rc::$Rc<T, dyn Destruct>: Send {}` which seems circular? Shouldn't rustc error?
* Even if somehow it's not circular, shouldn't having `NonNull` as a field make it so that neither `Arc` or `Rc` are Send nor Sync, so that these impls do nothing?
* Is it intentional that `$Rc` and `rc::$Rc` are both used? Since this is a private macro just used internally to create the types I'm not sure it matters.

I'm still fairly new to Rust so sorry if these are newbish questions.

**Comment by CAD97** [comment]
> We don't want any `Send`/`Sync` impl for `Rc` at all AFAICT, only `Arc`.

By gating the impls on std's impl, this allows the impl to be uniformly predicated for both cases, rather than pulled out separately. This is essentially writing explicitly what the automatic/inherited implementation would be if the implementation didn't require the use of raw pointers.

`where Arc<Owner>: Send + &T: Send` is equivalent to `where Owner: Send + Sync + T: Sync`; it's just a matter of which form is clearer. I have a weak preference for the "by example" form (it's theoretically harder to get wrong, and more directly states the actual requirements), but I do agree that impossible `Send`/`Sync` impls with unsatisfiable `where` bounds for `Rc` showing up in documentation isn't ideal.

> This seems to be saying that `Rc<T, Owner>` is `Send`+`Sync` if `Rc<T>` is. So `Owner` doesn't affect sendability/syncability, but I don't see why it couldn't? Is that what is improper?

Yes, that's what the impl says and what's wrong with it. The correct strucutral impl is bound on `Rc<Owner>` and `&T`, the two things that the mapped Rc logically contains and provides access to.

> Why does rustc accept this ... Is it intentional that `$Rc` and `rc::$Rc`

Note that `$Rc` is the local type, but `rc::$Rc` is the path to std's `$Rc`. This isn't the *clearest* way of writing it, but it allows reusing the same macro-generated implementation for both `Rc` and `Arc` rather than requiring writing/copying/editing it twice in source.

**Comment by CAD97** [comment]
> I'm also interested in tweaking the signature of project to return `U + '_` instead of `&U`.

This is far from a simple task. The key problem is that we can't store the field reference directly; we need to erase it somehow. For `&U` we can erase the lifetime and use `ptr::NonNull<U>` easily. I have a separate crate, [erasable](https://docs.rs/erasable/latest/erasable/), which defines an `ErasablePtr` trait for types which can be temporarily erased to an untyped `ptr::NonNull<()>`, but doing that means only supporting sized pointees, because we're not capable of storing a fat pointer.

If you want something more flexible than shared-rc, you probably want [yoke](https://docs.rs/yoke/latest/yoke/). `shared_rc::Rc<Field, Owner>` is `Yoke<&'static Field, Rc<Owner>>`. The reduction in complexity allows the shared-rc to be simpler than `Yoke`'s, but yoke offers essentially full flexibility. Personally, I don't think there's a middle-ground between shared-rc and yoke which is useful.

**Comment by CAD97** [comment]
I published version 0.1.1 which fixes the soundness issue and relaxes the `Clone` implementations to be actually useful. I also yanked the unsound version 0.1.0.

I hope to do a larger pass at improving the repo organization (e.g. add CI) eventually, along with moving the various [pointer-utils](https://github.com/CAD97/pointer-utils) crates into the point-rs organization. It's just hard to find time, unfortunately. I have a bigger improvement planned where I issue a breaking change for erasable once some amount of the `ptr_metadata` API stabilizes and unblocks the significant improvements relying on that, and motivating spending time before that becomes a possibility is difficult.
